### PR TITLE
SBND OpT0Finder Updates

### PR DESCRIPTION
**Pull request by lynnt-uchicago**
SBND OpT0Finder integration

dependencies on [sbncode PR #360](https://github.com/SBNSoftware/sbncode/pull/360) and [sbnobj PR #90](https://github.com/SBNSoftware/sbnobj/pull/90).

major updates: 
- turning _off_ Minuit minimization. This version of OpT0Finder in SBND is optimized only for one-to-many matching (one flash to many slices). 
- using`anab::Calorimetry` objects for track spacepoint conversion (charge-to-light)
- tuned shower constant conversion (charge-to-light)
- ~~can specify channels to mask~~ Masking is now on a match-specific basis instead of an entire event basis; masks are set for each flash and each slice, so the mask for the match is a combination of the flash-mask and slice-mask. 
- added function to use XARAPUCA flashes (default is PMT only) 
- default scoring method is to use normalized chi-squared mode 

more details in [docdb 30643](https://sbn-docdb.fnal.gov/cgi-bin/sso/ShowDocument?docid=30643).

### CODE REVIEW

**Comment by lynnt-uchicago** [comment]
@marcodeltutto I know it's been quite a while since you worked on any OpT0Finder code, but wanted to add you as a reviewer just in case. Let me know if you think it's appropriate to also add you to the corresponding PR for `sbncode`! Thank you :)

**Comment by marcodeltutto** [comment]
Hi @lynnt-uchicago! I'll be happy to review this and the one on `sbncode`. It will just take me a couple of days to get to it, hope that's ok!

File: sbndcode/OpT0Finder/SBNDOpT0Finder_module.cc [hunk 1]
```diff
@@ -429,76 +631,301 @@ bool SBNDOpT0Finder::ConstructLightClusters(art::Event& e, unsigned int tpc) {
...
+          auto track = track_v[n_trk];
+
+          // ** exiting track section ** 
+          // find if the track is uncontained and intersects the wire planes
+          bool uncontained = false;
+          auto const trk_start = track->Start();
+          auto const trk_end   = track->End();
+          geo::Point_t exit_pt; 
+
+          if (abs(trk_start.X()) >= 198.0) {exit_pt = trk_start; uncontained = true;}
```
Comment (marcodeltutto) [inline sbndcode/OpT0Finder/SBNDOpT0Finder_module.cc:706]: We should take the 198 from the detector properties service HalfWidth, but the TPC size is not going to change, so probably it's ok!

**Review by marcodeltutto — APPROVED** [review]

File: sbndcode/OpT0Finder/SBNDOpT0Finder_module.cc [hunk 2]
```diff
@@ -429,76 +631,301 @@ bool SBNDOpT0Finder::ConstructLightClusters(art::Event& e, unsigned int tpc) {
...
+                                          << exit_pt.Z() << std::endl;
+
+            int tpc = (exit_pt.X() > 0)? 1 : 0; 
+            for (size_t opch=0; opch < geo->NOpDets(); opch++){
+              if (int(opch)%2 != tpc) continue;
+              // only coated PMTs and vuv arapucas will be affected by direct light
+              if (_pds_map.isPDType(opch, "pmt_uncoated") || _pds_map.isPDType(opch, "xarapuca_vis")) continue;
+              if (_use_arapucas && _pds_map.isPDType(opch, "xarapuca_vuv")) continue;
+              auto center = _opch_centers.at(opch);
+              if ((abs(center.Z() - (exit_pt.Z() + 50*std::cos(track->Theta()))) <= 75) &&
```
[the remark this answers is not in the document]
Comment (lynnt-uchicago) [reply sbndcode/OpT0Finder/SBNDOpT0Finder_module.cc:722]: Will add comment! line 722 calculates whether or not the center of an optical detector is within a certain range of an exiting track's z exit point, with some leeway depending on the angle of the track (theta_xz). I could make the numbers 50 and 70 fcl configurable, which determines how much "leeway" you want to give.

**Comment by lynnt-uchicago** [comment]
trigger build LArSoft/lar*@LARSOFT_SUITE_v09_75_02 SBNSoftware/sbncode@v09_75_02#360 SBNSoftware/sbnanaobj@v09_21_02 SBNSoftware/sbnobj@v09_17_03#90

**Comment by RachelCoackley** [comment]
Sorry for the delay on testing this! I was waiting for the new sbndcode version to come out as I was having a few problems with getting the CI to use the other PRs and the correct versions (which is why there were build failures above). I've updated reference files this morning and ran this PR (wiith the dependencies) through the CI again, it got through all of the stages with just some warnings at the CI tests stage which are linked [https://dbweb8.fnal.gov:8443/LarCI/app/ns:sbnd/build_detail/phase_details?build_id=sbnd_ci/12293&platform=Linux%203.10.0-1160.88.1.el7.x86_64&phase=ci_tests&buildtype=slf7%20e20:prof](here)

**Comment by lynnt-uchicago** [comment]
Thanks Rachel, and no worries at all, appreciate your help!! The warnings that pop up are expected; I changed a lot of the fhicl structure and parameters ( which caused `(Problem: Differences in fhicl dump outputs)` and also created new data products (switched from using `anab::T0` to `sbn::OpT0Finder`, leading to `(Problem: Differences in products names *** Differences in products sizes ***`) . 

As for this warning: `Scaled cpu usage 246.13362450000002 not in 900.0:1500.0 range)`, I'm not super familiar with how the CI tests evaluate whether cpu usage is too high or too low, but compared to the previous OpT0Finder, in this feature branch the CPU is expected to be much much lower (in reco2 overall) and the stage to run faster. 

If the CI test reco2 stage has always included opt0finder, we should expect the reco2 stage cpu usage to be lower with this branch.

**Comment by lynnt-uchicago** [comment]
Hopefully last commit; Speaking of fhicl dump outputs, I forgot that for the reco_sce we do need to update the input producer labels for OpT0Finder (since now we access track/shower/calorimetry objects).

**Comment by RachelCoackley** [comment]
Great, thanks Lynn for going through all of the warnings! It all sounds good to me, the CPU usage is based on some limits that we set based on what we usually see when running the CI normally so it's not worrying that they have gone down

**Comment by lynnt-uchicago** [comment]
trigger build LArSoft/lar*@LARSOFT_SUITE_v09_75_03 https://github.com/SBNSoftware/sbncode/pull/360 https://github.com/SBNSoftware/sbnobj/pull/90

**Comment by fjnicolas** [comment]
[CI test ](https://dbweb8.fnal.gov:8443/LarCI/app/ns:SBND/build_detail/phase_details?build_id=sbnd_ci/12447&platform=Linux%203.10.0-1160.88.1.el7.x86_64&phase=ci_tests&buildtype=slf7%20e20:prof)successful in the last build of the parent PR (SBNSoftware/sbncode#360)
- Some CI warnings related to this PR previously advertised by Lynn
- Differences in the CAF files unrrelated to this PR ( due to SBNSoftware/sbncode#374)

**Pull request merged**